docs: Add guide for using `ClpKeyValuePairStreamHandler`. (#56)
y-scope/clp-loglib-py: README.md

can be installed with `pip`:

`python3 -m pip install --upgrade clp-logging`
  
## Logging handlers

### ClpKeyValuePairStreamHandler

⭐ *New in v0.0.14*

This handler enables applications to write structured log events directly into CLP's key-value pair
(kv-pair) IR stream format. The handler accepts structured log events in the form of Python
dictionaries, where each dictionary entry must abide by the requirements detailed
[below](#key-value-pair-requirements). The handler will also automatically include certain

> [metadata](#automatically-generated-kv-pairs), like the log event's level, with each log event.

Author: ```suggestion
[metadata](#automatically-generated-kv-pairs) (e.g., the log event's level) with each log event.
```

> [!NOTE]
> Since this handler accepts structured log events, it doesn't support setting a

> > [Formatter][py-logging-formatter] (because the log events don't need formatting into a string).

Author: ```suggestion
> [Formatter][py-logging-formatter] (because the log events don't need to be formatted into a
> string).
```
"to be formatted into a string"?

- Values must be one of the following types:
  - Primitives: `int`, `float`, `str`, `bool`, or `None`.
  - Arrays (`list`), where each array:
    - may contain primitive values, dictionaries, or nested arrays.
    - can be empty.
  - Dictionaries (`dict`), where each dictionary:
    - must adhere to the aforementioned rules for keys and values.
    - can be empty.

#### Automatically generated kv-pairs

In addition to the kv-pairs explicitly logged by the application, the handler will add kv-pairs,
like the log event's level, to each log event. We refer to the former as *user-generated* kv-pairs
and the latter as *auto-generated* kv-pairs.

> [!NOTE]
> The kv-pair IR stream format stores auto-generated kv-pairs separately from user-generated
> kv-pairs, so users don't need to worry about key collisions with the auto-generated keys.

The handler adds the following auto-generated kv-pairs to each log event:

| Key                 | Value type | Description                                       |
|---------------------|------------|---------------------------------------------------|
| `timestamp`         | `dict`     | The log event's timestamp                         |

> | - `unix_millisecs`  | `int`      | The timestamp as a Unix timestamp in milliseconds |

Author: ```suggestion
| - `unix_millisecs`  | `int`      | The timestamp in milliseconds since the Unix epoch |
```
Maybe this is less repetitive?

> | - `utc_offset_secs` | `int`      | The timestamp's UTC offset in seconds             |

Author: In my previous writing, I added an explanation about the direction of the offset. Do you think we can drop it safely? I'm not sure if there's any commonly used convention on it...
btw that explanation was directly copied from Python standard `time` library.

Reviewer: I think behind/ahead can be confusing when talking about time whereas an offset from a value should be clear? If you wan to be more explicit, we could say "The timestamp's offset fom UTC, in seconds"?

Author: Sure, that sounds better

Python's standard logging APIs. It operates similarly to Python's standard `logging.StreamHandler`,
with the following differences:
- Log events (`logging.LogRecord`) should contain the key-value pairs that a user wants to log
  as a Python dictionary.
  - These key-value pairs are written directly, without being formatted as a string.
- The key-value pairs will be serialized into the CLP key-value pair IR format before writing to
  the stream.

> [!NOTE]  
> The current `ClpKeyValuePairStreamHandler` does not support `CLPLogLevelTimeout`. This feature
> will be added in a future release.

### Key-value pairs as Python dictionary

Key-value pairs in the log event, presented as Python dictionaries, must abide by the following
rules:

- Keys must be of type `str`.
- Values must be one of the following types:
  - Primitives: `int`, `float`, `str`, `bool`, or `None`.
  - Arrays (`list`), where each array:
    - may contain primitive values, dictionaries, or nested arrays.
    - can be empty.
  - Dictionaries (`dict`), where each dictionary:
    - must adhere to the aforementioned rules for keys and values.

>     - can be empty.

Author: Some parts are directly copied from `ClpKeyValuePairStreamHandler`'s docstring.

  - "source_location" (`dict`):
    - "path" (`str`): Source file path of the logging statement.
    - "line" (`int`): Line number of the logging statement.
- **User-generated kv-pairs**: Key-value pairs provided by the application via the Python logging
  API. These are passed through without modification and serialized as-is.

### Example: Use `ClpKeyValuePairStreamHandler` to log Python dictionary

```python
import logging
from pathlib import Path
from clp_logging.handlers import ClpKeyValuePairStreamHandler

clp_handler = ClpKeyValuePairStreamHandler(open(Path("example.clp.zst"), "wb"))
logger: logging.Logger = logging.getLogger(__name__)
logger.addHandler(clp_handler)

logger.info({
    "message": "This is an example message",
    "machine_info": {
        "uid": 12345,
        "ip": "127.0.0.1",
    },
})
```

### Read key-value pair IR streams

The following options are available for reading and deserializing kv-pair IR streams generated by
this handler:
- [clp-ffi-py][clp-ffi-py-pypi]: This library provides [Deserializer][clp-ffi-py-deserializer-doc]
  to access a kv-pair IR stream in Python. See [this example][clp-ffi-py-deserializer-example] for
  usage details.

> - clp-ffi-js: TODO

Author: Do we want to mention log-viewer here?

>   

Author: We could also add a sentence to say the generated IR stream can be directly ingested into clp-s.
